lib/pos_blockchain.go

```diff
@@ -363,18 +360,15 @@ func (bc *Blockchain) validateTimeoutQC(desoBlock *MsgDeSoBlock, validatorSet []
 	return errors.New("IMPLEMENT ME")
 }
 
-// validateAncestorsExist checks that all ancestors of this block exist in the block index.
-// If an ancestor is not found, we'll return the block hash of the missing ancestor so the
-// caller can request this block.
-func (bc *Blockchain) validateAncestorsExist(desoBlock *MsgDeSoBlock) (_missingBlockHash *BlockHash, _err error) {
-	// Notes: starting from the block passed in, we'll look for the parent in the block index.
-	// 1. If the parent does not appear in the block index, we'll return the parent's hash.
-	// 2. If the parent exists in the block index AND is in the best chain, we can safely assume
-	//    that all ancestors exist in the block index.
-	// 3. If the parent exists in the block index but is not in the best chain, we repeat from
-	//    step 1 with the parent as the block passed in.
-	// TODO: Implement me
-	return nil, errors.New("IMPLEMENT ME")
+// validateAncestorsExist checks the parent exists in the block index. If it does not,
+// return the parent's hash. Else, return nil.
+func (bc *Blockchain) validateAncestorsExist(parentHash BlockHash) *BlockHash {
+	// check for parent in block index
+	// if parent does not exist in block index, return parent's hash
+	if _, exists := bc.blockIndex[parentHash]; !exists {
+		return &parentHash
+	}
+	return nil
```

Comment:
Just a nit. It's a bit strange to return the input block hash if it doesn't exist. You may as well make this function return a bool that signifies whether the the ancestors exist or not. This way, the naming and intended use of the function become dead simple.

```suggestion
// hasKnownAncestors checks the parent exists in the block index. It returns true if all ancestors
// of the provided block hash exist in the block index. It returns false otherwise.
func (bc *Blockchain) hasKnownAncestors(blockHash BlockHash) bool {
        _, exists := bc.blockIndex[blockHash]
        return exists
```